app/src/main/java/com/asfoundation/wallet/logging/send_logs/SendLogsRepository.kt

```diff
@@ -0,0 +1,134 @@
+    private val cacheDir: File,
+) {
+
+  private val sendStateBehaviorSubject = BehaviorSubject.createDefault(SendState.UNINITIALIZED)
+
+  fun saveLog(tag: String?, data: String): Completable {
+    return Completable.fromAction { logsDao.saveLog(LogEntity(null, tag = tag, data = data)) }
+        .subscribeOn(rxSchedulers.io)
+  }
+
+  fun updateLogs(): Maybe<File> {
+
+    return logsDao.updateLogs()
+        .andThen(
+            logsDao.getSendingLogs()
+                .flatMapMaybe { logs ->
+                  if(logs.isEmpty())
+                    return@flatMapMaybe Maybe.empty()
+
+                  val logsFile = File.createTempFile("log", null, cacheDir)
+                  val logContent = StringBuilder()
+
+                  logs.forEach { logEntity ->
+                    logContent.append(logEntity.created.toString())
+                    logEntity.tag?.let { logContent.append(" " + logEntity.tag) }
+                    logContent.appendLine(": " + logEntity.data)
+
+                    logsFile.appendText(logContent.toString())
+                    logContent.clear()
+                  }
+
+                  return@flatMapMaybe Maybe.just(logsFile)
+                }
+        )
+        .subscribeOn(rxSchedulers.io)
+  }
+
+
+  fun sendLogs(ewt: String): Completable {
+
+    return Single.zip(sendLogsApi.getSendLogsUrl(ewt), updateLogs().materialize(),
+        { awsInfo, materializedFile -> Pair(awsInfo, materializedFile) }
+    )
+        .flatMapCompletable {
+          val file = it.second.value ?: return@flatMapCompletable Completable.complete()
+
+          val filePart = MultipartBody.Part.createFormData("file", file.name,
+              file.asRequestBody("text/*".toMediaType()))
+
+          awsUploadFilesApi.uploadFile(
+              it.first.url,
+              it.first.fields.awsAccessKeyId.toRequestBody("text/plain".toMediaType()),
+              it.first.fields.signature.toRequestBody("text/plain".toMediaType()),
+              it.first.fields.policy.toRequestBody("text/plain".toMediaType()),
+              it.first.fields.key.toRequestBody("text/plain".toMediaType()),
+              filePart,
+          )
+        }
+        .andThen(
+            logsDao.deleteSentLogs()
+        )
+        .doOnComplete {
+          sendStateBehaviorSubject.onNext(SendState.SENT)
+        }
+        .onErrorComplete {
+          sendStateBehaviorSubject.onNext(SendState.ERROR)
+          return@onErrorComplete true
+        }
+
+        .doOnSubscribe {
+          sendStateBehaviorSubject.onNext(SendState.SENDING)
+        }
+        .subscribeOn(rxSchedulers.io)
+  }
+
+  private fun canLog(address: String): Single<Boolean> {
+    return sendLogsApi.getCanSendLogs(address)
+        .map { response: CanLogResponse -> response.logging }
+        .onErrorReturnItem(false)
+        .subscribeOn(rxSchedulers.io)
+  }
+
+  fun getSendLogsState(address: String): Observable<SendLogsState> {
```

Comment:
Same thing, rename to observeSendLogsState to be more explicit

Reply:
agreed, altered